### PR TITLE
Adaptions run control

### PR DESCRIPTION
**Pull request by SimonRubenDrauz**
adaptions in run_control in order to increase the recyclability for pandapipes and coupled nets. 
- Renaming initial_powerflow to initial_run
- Moving the raised errors in run_control to ctrl_variables, as these are powerflow-specific.
- splitting run_control in net_initialization, control_intialization, control_implementation, control_finalization
- splitting control_implementation into _control_step and _control_repair
- bugfix: check_final_convergence should be checked in each level, not just as final step

### CODE REVIEW

File: pandapower/control/run_control.py [hunk 1]
```diff
@@ -63,40 +63,50 @@ def check_for_initial_powerflow(controllers):
...
+def check_final_convergence(run_count, max_iter, errors, net_convergence):
     if run_count > max_iter:
         raise ControllerNotConverged("Maximum number of iterations per controller is reached. "
                                      "Some controller did not converge after %i power flows!"
                                      % run_count)
-    if not net["converged"]:
-        raise pp.LoadflowNotConverged("Controller did not converge because Loadflow did not"
-                                      "converge!")
+    if not net_convergence:
+        raise errors[0]("Controller did not converge because the grid did not converge!")
```
[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapower/control/run_control.py:109]: Yes, okay, I agree. As one can not expect to know what a pipeflow is when just installing pandapower, I would suggest not to mention pipeflow and stick to your first recommendation!

File: pandapower/control/run_control.py [hunk 2]
```diff
@@ -63,40 +63,50 @@ def check_for_initial_powerflow(controllers):
...
-        ctrl_variables = dict()
-        ctrl_variables["level"], ctrl_variables["controller_order"] = get_controller_order(net)
-        ctrl_variables["run"] = pp.runpp
-        ctrl_variables["initial_powerflow"] = check_for_initial_powerflow(
-            ctrl_variables["controller_order"])
+        ctrl_variables = ctrl_variables_default(net)
+
+    ctrl_variables["errors"] = (LoadflowNotConverged, OPFNotConverged)
+
+    kwargs["recycle"], kwargs["only_v_results"] = get_recycle(ctrl_variables)
```
[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapower/control/run_control.py:98]: To be honest: I was expecting that you won't like it :D. I try to figure out another solution!

File: pandapower/control/run_control.py [hunk 3]
```diff
@@ -63,40 +63,50 @@ def check_for_initial_powerflow(controllers):
...
+
+
+def prepare_run_ctrl(net, ctrl_variables, **kwargs):
     """
     Prepares run control functions. Internal variables needed:
 
     **controller_order** (list) - Order in which controllers in net.controller will be called
     **runpp** (function) - the runpp function (for time series a faster version is possible)
-    **initial_powerflow** (bool) - some controllers need an initial powerflow prior to the control step
+    **initial_run** (bool) - some controllers need an initial run of the powerflow prior to the control step
```
[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapower/control/run_control.py:89]: In pandapipes I just have to functions: run_control and prepare_run_control. Therefore, I will have my own docstrings!

[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapower/control/run_control.py:89]: grafik

looks like this. I would adapt the docstring accordingly

File: pandapower/control/run_control.py [hunk 4]
```diff
@@ -63,40 +63,50 @@ def check_for_initial_powerflow(controllers):
 
     for order in controllers:
         for ctrl in order:
-            if ctrl.initial_powerflow:
+            if hasattr(ctrl, 'initial_powerflow'):
+                ctrl.initial_run = ctrl.initial_powerflow
+                del ctrl.initial_powerflow
```
[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapower/control/run_control.py:68]: I see. But how can we make sure, that if some one is loading an old net can still do his/her analysis. Would you as an alternative rather allow both initial_run and initial_powerflow in each if-condition instead?

[the remark this answers is not in the document]
Comment (FlorianShepherd) [reply pandapower/control/run_control.py:68]: I would just not delete it and mention that it deprecated. After some time we can then delete the old parameter I'd say.

File: pandapower/control/run_control.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapower/control/run_control.py:90]: changed it and realized your recommendations. Do you agree with the changes?

File: pandapower/control/run_control.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapower/control/run_control.py]: I changed it back to the solution you implemented before. Do you agree now?

File: pandapower/control/run_control.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapower/control/run_control.py]: changed the comment accordingly

File: pandapower/control/run_control.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapower/control/run_control.py]: changed it and realized your recommendations. Do you agree with the changes?

[the remark this answers is not in the document]
Comment (FlorianShepherd) [reply pandapower/control/run_control.py]: But you delete it now or am I mistaken? I'd not delete it and mention that is should not be used with a logger.warning("Deprecation warning bla") or something like this.

[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapower/control/run_control.py]: no, actually not, if you check 'files changed' you find now following sollution:
grafik. Your reference is outdated.

**Pull request merged**